src/CommunityToolkit.Aspire.Hosting.PowerShell/PowerShellScriptResource.cs

```diff
@@ -0,0 +1,269 @@
+                                new( "PSInvocationState", args.InvocationStateInfo.State.ToString() ),
+                                new( "Reason", args.InvocationStateInfo.Reason?.Message ?? string.Empty ),
+                            ]
+                        };
+
+                        if (knownState == KnownResourceStates.Running)
+                        {
+                            state = state with
+                            {
+                                StartTimeStamp = DateTime.Now,
+                            };
+                        }
+
+                        if (KnownResourceStates.TerminalStates.Contains(knownState))
+                        {
+                            state = state with
+                            {
+                                StopTimeStamp = DateTime.Now,
+                            };
+                        }
+
+                        return state;
+                    });
+            };
+
+            if (this.TryGetLastAnnotation<PowerShellScriptArgsAnnotation>(out var scriptArgsAnnotation))
+            {
+                foreach (var scriptArg in scriptArgsAnnotation.Args)
+                {
+                    if (scriptArg is IValueProvider valueProvider)
+                    {
+                        var value = await valueProvider.GetValueAsync(cancellationToken);
+                        _ps.AddArgument(value);
+                    }
+                    else
+                    {
+                        _ps.AddArgument(scriptArg);
+                    }
+                }
+            }
+
+            try
+            {
+                _ = await _ps.InvokeAsync(_emptyInput, _output);
+            }
+            catch (PipelineStoppedException)
+            {
+                // This is expected when the pipeline is stopped (i.e. via BreakAsync), so we can ignore it.
+                // The pipeline will be stopped when the PowerShell instance is disposed.
+            }
+            catch (Exception ex)
+            {
+                scriptLogger.LogError(ex, "Error invoking PowerShell script: {Message}", ex.Message);
+                throw;
+            }
+        }
+
+        private void ConfigurePSDataStreams(ILogger logger, ResourceNotificationService notifications)
+        {
+            _output.DataAdded += (_, args) =>
+            {
+                var psObject = _output[args.Index];
+                logger.LogInformation("Output: {Output}", psObject.ToString());
+            };
+            _output.Completed += async (_, _) =>
+            {
+                await notifications.PublishUpdateAsync(this, state => state with
+                {
+                    State = KnownResourceStates.Finished,
+                    StopTimeStamp = DateTime.Now,
+                    ExitCode = _ps.HadErrors ? 1 : 0,
+                });
+
+                logger.LogInformation("Output completed");
+            };
```

Comment:
If you invert the order of the methods in here you can return the call to `PublishUpdateAsync` and skip the async/await usage.

Reply:
No, I don't think it's explicit guidance from MS, more a preference on my part. When methods are really just doing a pass-through to another async method, I prefer the simplicity of not having extra state machines created.

But looking at this one again, it's probably needed, because I'm not sure if you can assign the lambda to the event handler without an async.